src/jvm.rs

```diff
@@ -125,6 +133,100 @@ fn get_or_default_init_jvm() -> crate::GlobalResult<JvmPtr> {
     }
 }
 
+/// Invoked as the body from a JNI native function when it is called by the JVM.
+/// Initializes the environment and invokes `op`. Converts the result into a java
+/// object and returns it. Caller should then return this to the JVM.
+///
+/// # Safety condition
+///
+/// Must be invoked as the entire body of a JNI native function, with
+/// `env` being the `EnvPtr` argument provided.
+pub unsafe fn native_function_returning_object<J, R>(
+    env: EnvPtr<'_>,
+    op: impl FnOnce() -> R,
+) -> jni_sys::jobject
+where
+    J: Upcast<crate::java::lang::Object> + Upcast<J>,
+    R: ToJavaImpl<J> + std::fmt::Debug,
+{
+    init_jvm_from_native_function(env);
+    let old_state = thread::attach_from_jni_callback(env);
+
+    let result = match std::panic::catch_unwind(AssertUnwindSafe(|| op())) {
+        Ok(result) => {
+            let mut jvm = Jvm(env);
+            let obj = result.to_java().execute_with(&mut jvm);
+            match obj {
+                Ok(Some(p)) => p.into_raw().as_ptr(),
+                Ok(None) => std::ptr::null_mut(),
+                Err(e) => todo!("convert from exception `{e:?}` into java exception"),
```

Comment:
The `todo!` (here an inside `rust_panic_to_java_exception`) will skip the `detach_from_jni_callback`, breaking the invariant of `STATE`. I wonder how bad that is. Maybe some safety condition is violated as well.

Reply:
A type that calls `detach_from_jni_callback` on `Drop` might be useful, instad of manually checking all possible unwind paths.

Reply:
Great suggestion! Done.